### ISSUE TITLE
panic: runtime error: index out of range

### ISSUE THREAD

**Issue by TimSimmons** [rackerlabs/slappy#39]
I saw this behavior under the following weird condition in my dev environment when I didn't have BIND9 running, but Powerdns instead (don't ask). But I switched over to the `slappy` buffer and saw the following (instead of logs):

```
DEBUG: 2016/02/01 19:24:48.495861 Received request from 127.0.0.1:50253 for testdomain34014550.com. opcode: 14 rrtype: 65282 rrclass: 65280
panic: runtime error: index out of range

goroutine 3550 [running]:
github.com/rackerlabs/slappy/slapdns.get_serial(0xc82000e120, 0x17, 0xc82000e6ea, 0xc, 0x80, 0x0, 0x0)
	/opt/golang/src/github.com/rackerlabs/slappy/slapdns/slapdns.go:427 +0x620
github.com/rackerlabs/slappy/slapdns.handle_create(0xc82000e120, 0x17, 0xff00ff02, 0xc8200881b0, 0x7f046cfde028, 0xc820102070, 0x1)
	/opt/golang/src/github.com/rackerlabs/slappy/slapdns/slapdns.go:119 +0xd1
github.com/rackerlabs/slappy/slapdns.Handle(0x7f046cfde028, 0xc820102070, 0xc820088000)
	/opt/golang/src/github.com/rackerlabs/slappy/slapdns/slapdns.go:63 +0x889
github.com/rackerlabs/slappy/vendor/github.com/miekg/dns.HandlerFunc.ServeDNS(0x891840, 0x7f046cfde028, 0xc820102070, 0xc820088000)
	/opt/golang/src/github.com/rackerlabs/slappy/vendor/github.com/miekg/dns/server.go:82 +0x3a
github.com/rackerlabs/slappy/vendor/github.com/miekg/dns.(*ServeMux).ServeDNS(0xc820010930, 0x7f046cfde028, 0xc820102070, 0xc820088000)
	/opt/golang/src/github.com/rackerlabs/slappy/vendor/github.com/miekg/dns/server.go:186 +0xbb
github.com/rackerlabs/slappy/vendor/github.com/miekg/dns.(*Server).serve(0xc8200e40d0, 0x7f046d85f028, 0xc82000b3e0, 0x7f046d8600c8, 0xc820010930, 0xc8201062d0, 0x28, 0x28, 0x0, 0x0, ...)
	/opt/golang/src/github.com/rackerlabs/slappy/vendor/github.com/miekg/dns/server.go:535 +0x7c1
created by github.com/rackerlabs/slappy/vendor/github.com/miekg/dns.(*Server).serveTCP
	/opt/golang/src/github.com/rackerlabs/slappy/vendor/github.com/miekg/dns/server.go:453 +0x49a

goroutine 1 [select, 166 minutes]:
main.listen()
	/opt/golang/src/github.com/rackerlabs/slappy/main.go:46 +0x66a
main.main()
	/opt/golang/src/github.com/rackerlabs/slappy/main.go:82 +0x403

goroutine 17 [syscall, 166 minutes, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1696 +0x1

goroutine 5 [syscall, 166 minutes]:
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x18
created by os/signal.init.1
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x37

goroutine 6 [select, 166 minutes, locked to thread]:
runtime.gopark(0x892368, 0xc82002a728, 0x7f8d78, 0x6, 0x42da18, 0x2)
	/usr/local/go/src/runtime/proc.go:185 +0x163
runtime.selectgoImpl(0xc82002a728, 0x0, 0x18)
	/usr/local/go/src/runtime/select.go:392 +0xa64
runtime.selectgo(0xc82002a728)
	/usr/local/go/src/runtime/select.go:212 +0x12
runtime.ensureSigM.func1()
	/usr/local/go/src/runtime/signal1_unix.go:227 +0x353
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1696 +0x1

goroutine 7 [chan receive, 166 minutes]:
github.com/rackerlabs/slappy/vendor/github.com/vharitonsky/iniflags.sighupHandler(0xc82001a480)
	/opt/golang/src/github.com/rackerlabs/slappy/vendor/github.com/vharitonsky/iniflags/iniflags.go:140 +0x4d
created by github.com/rackerlabs/slappy/vendor/github.com/vharitonsky/iniflags.Parse
	/opt/golang/src/github.com/rackerlabs/slappy/vendor/github.com/vharitonsky/iniflags/iniflags.go:68 +0x29b

goroutine 9 [sleep]:
time.Sleep(0x1312d00)
	/usr/local/go/src/runtime/time.go:59 +0xf9
github.com/rackerlabs/slappy/stats.stats_listener()
	/opt/golang/src/github.com/rackerlabs/slappy/stats/stats.go:116 +0x2b5
created by github.com/rackerlabs/slappy/stats.Init_stats
	/opt/golang/src/github.com/rackerlabs/slappy/stats/stats.go:124 +0x90

goroutine 10 [chan receive]:
github.com/rackerlabs/slappy/stats.status_file()
	/opt/golang/src/github.com/rackerlabs/slappy/stats/stats.go:41 +0x2f8
created by github.com/rackerlabs/slappy/stats.Init_stats
	/opt/golang/src/github.com/rackerlabs/slappy/stats/stats.go:125 +0xa8

goroutine 11 [IO wait]:
net.runtime_pollWait(0x7f046d85ffa0, 0x72, 0xc8200101a0)
	/usr/local/go/src/runtime/netpoll.go:157 +0x60
net.(*pollDesc).Wait(0xc820102060, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc820102060, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).accept(0xc820102000, 0x0, 0x7f046d8600f0, 0xc82000eac0)
	/usr/local/go/src/net/fd_unix.go:408 +0x27c
net.(*TCPListener).AcceptTCP(0xc8200d8008, 0x891888, 0x0, 0x0)
	/usr/local/go/src/net/tcpsock_posix.go:254 +0x4d
github.com/rackerlabs/slappy/vendor/github.com/miekg/dns.(*Server).serveTCP(0xc8200e40d0, 0xc8200d8008, 0x0, 0x0)
	/opt/golang/src/github.com/rackerlabs/slappy/vendor/github.com/miekg/dns/server.go:435 +0x1c2
github.com/rackerlabs/slappy/vendor/github.com/miekg/dns.(*Server).ListenAndServe(0xc8200e40d0, 0x0, 0x0)
	/opt/golang/src/github.com/rackerlabs/slappy/vendor/github.com/miekg/dns/server.go:309 +0x3ec
main.serve(0x7f9178, 0x3, 0x0, 0x0, 0x7f4d78, 0x4)
	/opt/golang/src/github.com/rackerlabs/slappy/main.go:30 +0x47a
created by main.main
	/opt/golang/src/github.com/rackerlabs/slappy/main.go:79 +0x39b

goroutine 12 [IO wait]:
net.runtime_pollWait(0x7f046d85fee0, 0x72, 0xc8200101a0)
	/usr/local/go/src/runtime/netpoll.go:157 +0x60
net.(*pollDesc).Wait(0xc82010a060, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc82010a060, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).readMsg(0xc82010a000, 0xc82039fc00, 0x200, 0x200, 0xc8201062a0, 0x28, 0x28, 0xffffffffffffffff, 0x0, 0x0, ...)
	/usr/local/go/src/net/fd_unix.go:286 +0x33d
net.(*UDPConn).ReadMsgUDP(0xc82010c000, 0xc82039fc00, 0x200, 0x200, 0xc8201062a0, 0x28, 0x28, 0x200, 0x6fa120, 0x1, ...)
	/usr/local/go/src/net/udpsock_posix.go:96 +0x171
github.com/rackerlabs/slappy/vendor/github.com/miekg/dns.ReadFromSessionUDP(0xc82010c000, 0xc82039fc00, 0x200, 0x200, 0x200, 0x200, 0x0, 0x0)
	/opt/golang/src/github.com/rackerlabs/slappy/vendor/github.com/miekg/dns/udp.go:47 +0xc1
github.com/rackerlabs/slappy/vendor/github.com/miekg/dns.(*Server).readUDP(0xc8201000d0, 0xc82010c000, 0x77359400, 0x0, 0x0, 0x0, 0xc820100188, 0x0, 0x0)
	/opt/golang/src/github.com/rackerlabs/slappy/vendor/github.com/miekg/dns/server.go:605 +0x190
github.com/rackerlabs/slappy/vendor/github.com/miekg/dns.(*defaultReader).ReadUDP(0xc82010c028, 0xc82010c000, 0x77359400, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/opt/golang/src/github.com/rackerlabs/slappy/vendor/github.com/miekg/dns/server.go:230 +0x5f
github.com/rackerlabs/slappy/vendor/github.com/miekg/dns.(*Server).serveUDP(0xc8201000d0, 0xc82010c000, 0x0, 0x0)
	/opt/golang/src/github.com/rackerlabs/slappy/vendor/github.com/miekg/dns/server.go:478 +0x1fe
github.com/rackerlabs/slappy/vendor/github.com/miekg/dns.(*Server).ListenAndServe(0xc8201000d0, 0x0, 0x0)
	/opt/golang/src/github.com/rackerlabs/slappy/vendor/github.com/miekg/dns/server.go:327 +0x6fc
main.serve(0x7f92e0, 0x3, 0x0, 0x0, 0x7f4d78, 0x4)
	/opt/golang/src/github.com/rackerlabs/slappy/main.go:30 +0x47a
created by main.main
	/opt/golang/src/github.com/rackerlabs/slappy/main.go:80 +0x3fe
```
This was pretty reproducible with slappy running on Debian Jessie and PowerDNS (`3.4.1`). The git ref for miekg/dns at the time was https://github.com/miekg/dns/tree/1756430e42a7b2ecded216a9fdd37d002c116df5

The panic occured here:
https://github.com/rackerlabs/slappy/blob/a2a295a41754bcbdfae5253798f0f7ff963bd2b3/slapdns/slapdns.go#L427

```
	return serial_query_parse(in), nil  // HERE
}

func serial_query_parse(in *dns.Msg) uint32 {
	var serial uint32 = 0
	if in.Rcode != dns.RcodeSuccess {
		return serial
	}
	if rr, ok := in.Answer[0].(*dns.SOA); ok {
		serial = rr.Serial
	}
	return serial
}
```

@pglass guessed that this function was being inlined by the Go compiler, and was confirmed in the #go-nuts IRC channel:
```
[14:29:29] pglass:	 does the go compiler inline function calls?
[14:29:58] Tv`:	 pglass: some, yes
[14:30:20] pglass:	 Tv`: do you know how it decides?
[14:30:34] Tv`:	 pglass: no guarantees ;)
[14:31:04] pglass:	 Tv`: well, is there a way to force a function to be inlined?
[14:31:07] Tv`:	 no
[14:36:56] pglass:	 Tv`: last question, is there an option to disable function inlining, so that functions aren't "missing" from stacktraces?
[14:37:51] Tv`:	 pglass: only as a big blunt hammer, -gcflags something to disable optimizations
[14:38:50] Tv`:	 https://golang.org/cmd/compile/
[14:40:01] pglass:	 ah, `go build -gcflags -m` will tell you when it inlines a function!
```

The function is being inlined, as confirmed by:
```
$ go build -gcflags '-l -m' main.go
...
slapdns/slapdns.go:430: can inline serial_query_parse
...
./main --config slappy.conf
...
DEBUG: 2016/02/01 20:47:00.000986 Received request from 127.0.0.1:54957 for testdomain29654075.com. opcode: 14 rrtype: 65282 rrclass: 65280
panic: runtime error: index out of range

goroutine 14 [running]:
github.com/rackerlabs/slappy/slapdns.serial_query_parse(0xc820100000, 0x0)
	/opt/golang/src/github.com/rackerlabs/slappy/slapdns/slapdns.go:435 +0x7b
```

The problem is happening at:
`if rr, ok := in.Answer[0].(*dns.SOA); ok {` because the version of PowerDNS that I have will give back a `NOERROR` Rcode (thus passing the above check) even if it has no answer section ( lol pdns http://blog.powerdns.com/2015/03/02/from-noerror-to-refused/). So when we try to index the empty answer section, boom.

Amazing spelunking @pglass !

tl;dr: If this agent needed to support PDNS `< 3.4.3`, it needs a change in the `serial_query_parse` function to look for empty answers rather than Rcodes.

**Comment by rsyed83** [comment]
That's some nice debugging there!

![](http://www.reactiongifs.us/wp-content/uploads/2013/03/slow_clap_citizen_kane.gif)
